File: Sources/WearScope/WSCore.swift
/*
 * WSCore — event buffering, persistence, batch upload.
 *
 * Design (docs/OBSERVABILITY.md §4):
 *  - Events append to a jsonl file immediately (crash survival)
 *  - Flush every 50 events or 10 s: POST /v1/ingest if an endpoint is set, else local mode (file only)
 *  - Truncate the file on upload success; keep and retry on failure. 2 MB file cap
 */

import Foundation

actor WSCore {
  static let shared = WSCore()

  private var config: WSConfig?
  private var sessionId = UUID()
  private var sessionStart = Date()
  private var pending: [WSEvent] = []
  private var flushTask: Task<Void, Never>?
  private var fileURL: URL?

  private let maxFileBytes = 2_000_000
  private let flushCount = 50
  private let flushSeconds: UInt64 = 10

  // MARK: - Lifecycle

  func start(config: WSConfig) {
    self.config = config
    sessionId = UUID()
    sessionStart = Date()

    let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
      .appendingPathComponent("WearScope", isDirectory: true)
    try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    fileURL = dir.appendingPathComponent("events.jsonl")

    add(WSEvent(type: .custom, name: "sdk_start", attrs: [
      "sdk": WSConfig.sdkVersion,
      "mode": config.endpoint == nil ? "local" : "cloud",
      "locale": Locale.current.identifier,
      "timezone": TimeZone.current.identifier,
    ]))

    flushTask?.cancel()
    flushTask = Task { [weak self] in
      while !Task.isCancelled {
        try? await Task.sleep(nanoseconds: (self?.flushSeconds ?? 10) * 1_000_000_000)
        await self?.flush()
      }
    }
  }

  var localFileURL: URL? { fileURL }

  // MARK: - Events

  func add(_ event: WSEvent) {
    guard config != nil else { return }
    pending.append(event)
    appendToFile(event)
    if pending.count >= flushCount {
      Task { await flush() }
    }
  }

  func flush() async {
    guard let config, config.endpoint != nil else {
      pending.removeAll()  // local mode: already written to the file
      return
    }
    guard !pending.isEmpty else { return }
    let batch = pending
    pending.removeAll()

    let envelope = WSEnvelope(
      sdk: .init(name: "wearscope-swift", version: WSConfig.sdkVersion),
      app: config.appInfo,
      device: config.deviceInfo,
      session: .init(id: sessionId, startedAt: sessionStart),
      events: batch)

    guard let url = config.endpoint?.appendingPathComponent("v1/ingest"),
          let body = try? JSONEncoder.gs.encode(envelope) else { return }

    var req = URLRequest(url: url)
    req.httpMethod = "POST"
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    req.setValue(config.apiKey, forHTTPHeaderField: "X-API-Key")
    req.httpBody = body

    do {
      let (_, resp) = try await URLSession.shared.data(for: req)
      if let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode) {
        truncateFile()
      } else {
        pending.insert(contentsOf: batch, at: 0)  // retry next cycle (still on file too)
      }
    } catch {
      pending.insert(contentsOf: batch, at: 0)
    }
  }

  // MARK: - File persistence

  private func appendToFile(_ event: WSEvent) {
    guard let fileURL, var line = try? JSONEncoder.gs.encode(event) else { return }
    line.append(0x0A)
    if let size = try? fileURL.resourceValues(forKeys: [.fileSizeKey]).fileSize,
       size > maxFileBytes {
      truncateFile()  // over the cap: drop the old data (simple policy — ring buffer in v0.2)
    }
    if let handle = try? FileHandle(forWritingTo: fileURL) {
      defer { try? handle.close() }
      _ = try? handle.seekToEnd()
      try? handle.write(contentsOf: line)
    } else {
      try? line.write(to: fileURL)
    }
  }

  private func truncateFile() {
    guard let fileURL else { return }
    try? Data().write(to: fileURL)
  }
}

File: android/wearscope/src/main/kotlin/io/wearscope/WSCore.kt
/*
 * WSCore — event buffering, persistence, batch upload (same design as the iOS WSCore).
 *
 *  - Events append to a jsonl file immediately (crash survival)
 *  - Flush every 50 events or 10 s: POST /v1/ingest if an endpoint is set, else local mode (file only)
 *  - Truncate the file on upload success; keep and retry on failure. 2 MB file cap
 *
 * Zero dependencies: HttpURLConnection + org.json + a single-thread executor.
 */
package io.wearscope

import android.content.Context
import android.os.Build
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.time.Instant
import java.util.UUID
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import org.json.JSONArray
import org.json.JSONObject

internal object WSCore {
  private const val MAX_FILE_BYTES = 2_000_000L
  private const val FLUSH_COUNT = 50
  private const val FLUSH_SECONDS = 10L

  private val executor = Executors.newSingleThreadScheduledExecutor { r ->
    Thread(r, "WearScope").apply { isDaemon = true }
  }

  private var apiKey = ""
  private var endpoint: String? = null
  private var sessionId = UUID.randomUUID().toString()
  private var sessionStart: Instant = Instant.now()
  private val pending = mutableListOf<WSEvent>()

  private var file: File? = null
  private var appJson = JSONObject()
  private var deviceJson = JSONObject()

  fun start(context: Context, apiKey: String, endpoint: String?) {
    this.apiKey = apiKey
    this.endpoint = endpoint?.trimEnd('/')
    sessionId = UUID.randomUUID().toString()
    sessionStart = Instant.now()

    val dir = File(context.filesDir, "wearscope").apply { mkdirs() }
    file = File(dir, "events.jsonl")

    val pm = context.packageManager
    val info = pm.getPackageInfo(context.packageName, 0)
    appJson = JSONObject()
        .put("bundleId", context.packageName)
        .put("version", info.versionName ?: "0")
        .put("build", info.longVersionCode.toString())
    deviceJson = JSONObject()
        .put("model", Build.MODEL)
        .put("os", "Android ${Build.VERSION.RELEASE}")

    executor.scheduleAtFixedRate({ flushNow() }, FLUSH_SECONDS, FLUSH_SECONDS, TimeUnit.SECONDS)
  }

  fun add(event: WSEvent) {
    executor.execute {
      pending.add(event)
      appendToFile(event)
      if (pending.size >= FLUSH_COUNT) flushNow()
    }
  }

  fun flush() {
    executor.execute { flushNow() }
  }

  fun exportFile(): File? = file

  /**
   * Switch a running local-mode session to cloud mode once credentials arrive
   * (zero-config provisioning). Events buffered so far upload on the next flush.
   */
  fun adopt(apiKey: String, endpoint: String) {
    executor.execute {
      this.apiKey = apiKey
      this.endpoint = endpoint.trimEnd('/')
    }
    add(WSEvent(WSEventType.CUSTOM, "provisioned", mapOf("mode" to "cloud")))
  }

  // MARK: - Internals (all on the executor thread only)

  private fun flushNow() {
    val ep = endpoint
    if (ep == null) {  // local mode: already written to the file
      pending.clear()
      return
    }
    if (pending.isEmpty()) return
    val batch = pending.toList()
    pending.clear()

    val envelope = JSONObject()
        .put("sdk", JSONObject().put("name", "wearscope-kotlin").put("version", WearScope.SDK_VERSION))
        .put("app", appJson)
        .put("device", deviceJson)
        .put("session", JSONObject().put("id", sessionId).put("startedAt", sessionStart.toString()))
        .put("events", JSONArray(batch.map { it.toJson() }))

    try {
      val conn = URL("$ep/v1/ingest").openConnection() as HttpURLConnection
      conn.requestMethod = "POST"
      conn.setRequestProperty("Content-Type", "application/json")
      conn.setRequestProperty("X-API-Key", apiKey)
      conn.doOutput = true
      conn.connectTimeout = 10_000
      conn.readTimeout = 10_000
      conn.outputStream.use { it.write(envelope.toString().toByteArray()) }
      val code = conn.responseCode
      conn.disconnect()
      if (code in 200..299) {
        truncateFile()
      } else {
        pending.addAll(0, batch)  // retry next cycle (still on file too)
      }
    } catch (_: Exception) {
      pending.addAll(0, batch)
    }
  }

  private fun appendToFile(event: WSEvent) {
    val f = file ?: return
    if (f.length() > MAX_FILE_BYTES) truncateFile()  // over the cap: drop the old data (simple policy)
    runCatching { f.appendText(event.toJson().toString() + "\n") }
  }

  private fun truncateFile() {
    runCatching { file?.writeText("") }
  }
}
